### ISSUE TITLE
Handle exceptions in registered global event handlers more gracefully

### ISSUE THREAD

**Issue by WikkidEdd** [microsoft/MixedRealityToolkit-Unity#7126]
## Describe the problem
We had a hard to track down bug occur in our app where it appeared that the input system shut down seemingly randomly. No hand visualisation and no pointers, so the app became unuseable and required a restart.

Fortunately we had 2 MRCs of the app from two occasions when the bug occurred. It appeared that the bug occurred when the user said "Select". After much digging through the MRTK it turns out that saying "Select" triggers an OS level input action which the MRTK picks up and responds to. When this happens the GazePointer triggers an OnPointerDown event.

The bug in my case was occurring because I have registered a GlobalHandler for OnPointerDown with the input system, but inside that event I was assuming that a pointer would always have Result, but because the the GazePointer was disabled when the event came in, this wasn't the case and it caused an exception.

The exception was clearly a bug in my code, but it was inconsequential. If the outcome hadn't been so catastrophic, we would have shipped the app without ever knowing about it or caring.

The reason why it was so catastrophic is to do with a section of code within the BaseEventSystem.HandleEvent which keeps track of the recursion depth of the function. In between `eventExecutionDepth++;` and `eventExecutionDepth--;` the register hanlders are invoked. If one of them throws and expection. `eventExecutionDepth--;` never gets called. Which prevents `eventExecutionDepth` ever getting to 0. This has knock ons through out the event system preventing new handles from being register and unregister amongst other things.

When this happened it really wasn't obvious what could cause this outcome because it doesn't manifest immediately. Any active hands/pointers at the time the exception occurs keep functioning until tracking is lost, then after they won't come back.

Had we not luckily had the MRCs it would have taken us days to track down the issue.

## Describe the solution you'd like

It feels like the MRTK should be able to handle exceptions in registered handlers without this outcome. The way that the exception's effects are delayed makes it particular tricky to track down. 

To prevent this happening again, we've just wrapped the appropriate bits of code in a try/catch block so `eventExecutionDepth--;` will always get called.

I'm not convinced that this is the best fix as it potentially hides exceptions. Maybe getting it to fail at the point of the exception would be better. Perhaps someone more familiar with that area of the code can suggested a better solution?

**Comment by MenelvagorMilsom** [comment]
Thanks for the detailed PR description. We'll look into this further.

**Comment by MenelvagorMilsom** [comment]
Assigning to @davidkline-ms for initial triage

**Comment by wiwei** [comment]
I took a peek at this one - the main instance of this was fixed up in this change:

https://github.com/microsoft/MixedRealityToolkit-Unity/commit/9314eea6516450f86c20543460de80026786f62f

It looks like this got merged in ~December so missed the last release, but should be out with the next one.

There's one more spot in the code where it's "possible" that an issue could exist, but because of our usage of ExecuteEvents.Execute (which internally has exception eating code) it shouldn't be an issue. Nevertheless, I can also safeguard that just to be sure.
